### ISSUE TITLE
Migration to vector tiles

### ISSUE THREAD

**Issue by kocio-pl** [gravitystorm/openstreetmap-carto#3201]
Follow up to https://github.com/gravitystorm/openstreetmap-carto/issues/1975#issuecomment-384858383

As discussed privately among team members, it's good to migrate osm-carto to be rendered with [vector tiles](https://wiki.openstreetmap.org/wiki/Vector_tiles). To allow as smooth and painless migration path as possible, it should be ready for server side raster tiles rendering. It should be non-disruptive change for clients while allowing to start developing client side rendering and testing vector tiles in general.

This is not a fresh and untested approach, @gravitystorm has been using it for his own styles for a few years already - you can watch his presentation about it from SotM EU 2014:

https://www.youtube.com/watch?v=NoLJHgqDCzc

There's also a vector fork of osm-carto by @rory (also his presentation from SotM 2016):

https://github.com/geofabrik/openstreetmap-carto-vector-tiles
https://2016.stateofthemap.org/2016/converting-a-regular-carto-project-to-vector-tiles-osmcarto-case-study/

However this fork is quite old and happened before our v4.x database migration. It's not clear now how useful it could be, other than being proof of concept.

This ticket is about coordination of the effort, more specific tickets might be needed for subtasks and technical background (like choosing the proper software stack) is yet to be discussed and tested.

**Comment by jbelien** [comment]
Would be really awesome to support Vector Tiles ! 🎉 
But imho I do not think we should migrate ... We should add Vector Tiles support while still providing Raster Tiles :)

**Comment by StyXman** [comment]
I would agree, but maintaining two different styles is too complicated, specially because they don't seem to be compatible, or at least nobody has written a converter from one to the other. Think about all the contributions, having to be implemented twice...

**Comment by kocio-pl** [comment]
This problem is much wider than style, it's also related to deployment strategy (especially on OSMF servers, but most probably also for the forks following us, like German or Belgian) and tools to be used.

@jbelien The whole point with server side raster tiles rendering is about planning deployment so that vector osm-carto tiles are then transformed into raster tiles. In other words instead of:

osm-carto -> raster tiles

it would be:

osm-carto -> vector tiles -> raster tiles

so the final output should be very similar. 

Please note however that from the osm-carto perspective it's just a migration. Raster tiles rendering will be done with some other tools, so it's a deployment problem. Yet my intention is to plan the whole process with our partners (including you :smile:), not just abrupt change of the styling output.

@StyXman Rory's vector fork is designed to be easily synchronized with standard version, but that might not be the proper solution. Andy suggested making a development branch form the current release (like v4.10.0-vector) and not try to synchronize them. After some time, when we will be ready to really deploy vector tiles on production, we should only update the vector fork to the current version of plain osm-carto (like v4.44.4) or just port it directly (into v4.44.4-vector), whatever will be easier. This would be the switch point. From now on the vector version (which will become v5.0.0 most probably - or we could rename the whole project) will take over and the raster version will become the legacy. 

So we plan to make only some plain development fork just to learn things, but after that there will be no more two versions, the old one won't be supported probably. There's always possibility that somebody will like to support it, but given that the migration should not affect end users (and possibly be friendly for our forks) and there are only few people involved, I don't quite believe it would be the case.

**Comment by kocio-pl** [comment]
This ticket is exactly to make the discussion more open. You're right that this is not anything set in stone, it's just a draft made by Andy (based on his real life experience) and approved by me, so we could start with some clear vision. Thanks for a disclaimer.

So yes - take it with a big grain of salt... and join the discussion!

> Personally and from a design perspective i think a fresh start as with the change from Mapnik XML to CartoCSS would - even if it maintains some level of design continuity - be a very healthy step.

It was hard enough in 2012 [for one determined man to make it](https://www.youtube.com/watch?v=RgSye4IbnD8). Now it's [much more complex](http://paulnorman.ca/blog/2015/11/openstreetmap-carto-complexity/), so I don't believe that somebody else will try to do that again. 

But if you - or anybody else for that matter - really want to make such rewrite, it can be just as good solution. It might be even better in the long run, given your cartographic experience together with designing and coding skills. I just think it will take years to come with a first demo after the "simple" port will be ready, so I want to start with more limited approach, which I find to be more realistic.

I entirely share your position that any kind of competition could only make the state of OSM rendering more healthy.

**Comment by StyXman** [comment]
A consequence of doing vector tiles is, I think, the ability to turn on/off on the client the rendering certain features, typically POIs. This would probably help with things like rendering offices and, maybe, with PIOs that fall in multiple categories (for instance, here in .fr, some bakeries also sell take away food and coffee). Just an idea; I really don't know if this is possible.

**Comment by andrzej-r** [comment]
If the same stylesheet code can support both deployment targets then I think we could (and should) reuse/extend the current project. Otherwise, I'm with @imagico on that - it is better to fork the project under a new name, so that we don't force existing users to switch, and let people use and maintain the existing implementation if they wish so. Otherwise we end up with Gnome3 scenario where Gnome2 users have to fork and copy the existing implementation.

And, to make it clear, I am in favour of using vector tiles on the main osm map. Especially if, as planned, it can be rendered on the server side to avoid Web-GL. Many of the current usability issues (lack of language versions, lack of overlays for transport, POI categories etc) could be easily solved.

**Comment by kocio-pl** [comment]
Sure, we could make a fork(s) under some other name. Instead of long and muddy "OSM Carto Classic/Legacy" I would rather go with something like "Vektor Flagship" and call the other styles/deployments from the same family like "Vektor Drakkar" or some other fancy vessel type that makes sense. Right name can be quite important to avoid confusion (see #2927) and with many possible styles basing on the same vector engine such a double part name could help identify them (hence "Vektor" instead of "Vector" - not too common name on the net).

It's an interesting idea that new style could be based on automatic conversion of existing code, but I don't believe it. It would mean writing code parser and this is not trivial programming task, as seen lately on a very simple example (see #3151).

That means that we have not too much options left. One is kind of Rory's fork, which tries to add some vector-related things on top of existing CartoCSS+SQL syntax and I believe it's the way to go. Another is to write something from scratch and this would be quite another fork, which basically would be very light at the beginning. It might be nice competitor, but it would not allow smooth upgrade. That's why I want to keep the Andy's path. It's easier to make a lighter skin for such full-featured fork than make a different fork and try to recreate osm-carto features.

**Comment by gravitystorm** [comment]
@kocio-pl Custom software, based on the tilelive modules.

**Comment by pnorman** [comment]
tileigi is not suitable for the osm.org because it only does pre-generation of vector tiles, which won't work with regular updates. It's also not really stable enough that I'd trust it in production yet.

**Comment by matthijsmelissen** [comment]
@pnorman I'm not sure if I follow you entirely. Why would polygon labels cause a difference between the current style and the vector port?

When we use ST_PointOnSurface, we have no guarantee about the location that is chosen (except that is lies in the polygon), right?

**Comment by kocio-pl** [comment]
Very good news! :+1: 

I wait for the summary of your talk with Rory. In case you wonder how it looks like, the new style is named Bolder and you can find it here: https://github.com/pnorman/bolder

**Comment by pnorman** [comment]
From the conversation with @rory, we discussed a few things

Geofabrik's interest is multiple languages with reasonable performance, retina tiles, and slightly changed styles for clients (colour changes, dropped features, etc). Vector tiles are good at this.

Rory found tilelive had bad performance, so he started tileigi, which can be 100x at generating tiles. Tileigi currently only pregenerates tiles, it doesn't serve them live. Most of the performance gains he got are from using vector metatiles: the practice of requesting the area of 8x8 tiles, and slicing the result up into 64 individual tiles, and storing individual tiles on disk.

osm-carto is a very complex style, and it can take hundreds of milliseconds to go from vector to raster for a single tile. This is exceptionally slow, and I'd consider times under 10ms more "normal". We discussed a number of ways to improve performance to make up for the osm-carto complexity and number of layers. The most likely route is to carry the metatiling farther. Instead of querying the area of 8x8 z14 tiles and slicing them before putting them on disk, put the unsliced tile (a z11 tile with a high resolution) on disk. Then, when a request for a z14 raster tile comes in, fetch the combined tile, generate a raster for the 8x8 area, and store the sliced tiles in memory. The in-memory tiles can be used for the subsequent requests for nearby tiles. This is technically overzoom for the vectors, and metatiles for the rasters.

What you can see from the above is that there is no software serving stack for Mapnik vector tiles that is 100% there for our needs.

-----

Some more changes that would help maintaining a vector fork are 

- Drop use of classes for layers. You can't do classes with vector tiles
- way_pixels is an issue for where overzoom is used. See rory's presentation for details, but we couldn't think of a solution for this.
- We could set buffers per-layer. Neither of us were sure what would happen to raster if buffers were set per-layer.
- #2524 is an issue with retina, which matters more when we have easy support for retina.

**Comment by kocio-pl** [comment]
From my point of view:
- rendering based on the size is essential, for example I was not able to find anything better for natural reserves to avoid mess; but since the area size data are in the database and the software is somehow aware what is the current zoom, so it should be possible to calculate current size of the area, or am i missing something?
- retina is just the nice addition
- I have no idea about what could be done with classes and buffers

But also:
- What about vector patterns (#2045)? Maybe just replacing the raster images with vector ones in the new implementation  would be enough.
- What about software support? Mapnik is supported and developed, so is Kosmtik and Carto CSS is on a way to become legacy, but still supported. What about new stack?
- Is there a chance to deploy this style in parallel on OSM.org? Pro: it can use the same database, contra: vector tiles can have many output formats (server side raster tiles, vector tiles with different looks) and I'm not sure how to serve them
- Is there a way to port the osm-carto features more easily than just rewriting them all from scratch by hand?

**Comment by dieterdreist** [comment]
2018-04-30 18:43 GMT+02:00 kocio-pl <notifications@github.com>:

> From my point of view:
>
>    - rendering based on the size is essential, for example I was not able
>    to find anything better for natural reserves to avoid mess; but since the
>    area size data are in the database and the software is somehow aware what
>    is the current zoom, so it should be possible to calculate current size of
>    the area, or am i missing something?
>
>

with vector tiles you already have this precomputed and filtered for area
labels. You only have a point and a size / importance class for it
(referring to the curent zoomlevel or range of levels), and you won't have
label nodes for vector tiles at zoom levels where you don't want to show
them. At least this is how mapbox does it.

**Comment by kocio-pl** [comment]
This is how Bolder looks like when rendered with repo instructions and Liechtenstein data:

![screenshot-2018-4-30 tegola mvt tile server](https://user-images.githubusercontent.com/5439713/39440631-c7c20944-4cab-11e8-8cf9-0ba449bc996b.png)

I was wondering about possible geometry distortion, which is a big issue for @imagico. Here it is on z19 compared to osm-carto on z19 - I don't understand why the size is different with the same zoom level, but at least geometry seems to be unchanged (even without "Turning Off Simplification of Geometries" option):

![screenshot-2018-4-30 tegola mvt tile server 2](https://user-images.githubusercontent.com/5439713/39440763-4e0b10a4-4cac-11e8-9e5d-e7167a87ece0.png)
![screenshot-2018-4-30 openstreetmap](https://user-images.githubusercontent.com/5439713/39440766-50f4f3fc-4cac-11e8-884f-262770041503.png)

**Comment by kocio-pl** [comment]
Bolder client-side style with Mazovian voivodship/Warsaw data:

![screenshot-2018-4-30 bolder - tangram js](https://user-images.githubusercontent.com/5439713/39442025-207edc20-4cb0-11e8-8582-f0d097b6bc84.png)
![screenshot-2018-4-30 bolder - tangram js 1](https://user-images.githubusercontent.com/5439713/39442026-22875074-4cb0-11e8-80e9-ad81cd51d11d.png)
![screenshot-2018-4-30 bolder - tangram js 2](https://user-images.githubusercontent.com/5439713/39442028-2462dcec-4cb0-11e8-8701-1baaa563a0f9.png)

**Comment by pnorman** [comment]
Client-side rendering is incompatible with the approach that Andy has suggested of incremental work, using small steps.

> software is somehow aware what is the current zoom,

afaik, Mapnik doesn't let you use the zoom in calculations. If it did, we could stop converting area to pixels, and do the math in the MSS. But without that ability, we need to use different thresholds for different levels of overzoom.

> retina is just the nice addition

It's one of the big reasons to use vector tiles. I don't know osm.org's usage numbers, but in general >50% of users can be expected to have a device with a high DPI screen.

> What about vector patterns (#2045)? Maybe just replacing the raster images with vector ones in the new implementation would be enough.

There's no direct relation with vector patterns. It's already an issue. It might become more common to do retina tiles with vector tiles, so more people might hit the blurry issue.

> What about software support? Mapnik is supported and developed, so is Kosmtik and Carto CSS is on a way to become legacy, but still supported. What about new stack?

We'd still be using Mapnik, Kosmtik, and CartoCSS. We'd need some Mapnik-based vector tile server for production, probably something based on tilelive. There are no clear candidates here. Tessera, Kartotherian, and others all would sort of work, but I'm not convinced that vector tiles have the performance without tricks like mentioned above, and no one has coded a stack with those kinds of tricks yet.

> Is there a way to port the osm-carto features more easily than just rewriting them all from scratch by hand?

There's no rewriting involved. See @rory's talk for what's needed to port the style.

**Comment by kocio-pl** [comment]
> afaik, Mapnik doesn't let you use the zoom in calculations. If it did, we could stop converting area to pixels, and do the math in the MSS. But without that ability, we need to use different thresholds for different levels of overzoom.

I see. @talaj Could you comment on this? Is this something doable in a near future or maybe something that could be ported from your fork, for example?

**Comment by talaj** [comment]
@kocio-pl This is possible already by [render-time variables](https://github.com/mapnik/mapnik/issues/2553) . `@zoom` variable [is implemented in tilelive-mapnik](https://github.com/mapbox/tilelive-mapnik/commit/62339a2f49a3449372a33a8f481bde9194030241).

**Comment by kocio-pl** [comment]
Sounds good, thanks!

> There's no rewriting involved. See @rory's talk for what's needed to port the style.

I will watch it once again, but I'm not sure what your project has to do with his effort. I can understand that vector tile production and serving might be done with different tools, but I don't understand how the CartoCSS rules can be used in Bolder?

**Comment by systemed** [comment]
Moving to vector tiles gives an opportunity to decouple cartography and schema, which I believe is essential.

Porting osm-carto to vector tiles as a single project would achieve some nice-to-haves (clickable POIs, better multilingual support, sharper display on retina imagery, better zooming) but nothing transformational.

Providing a style-agnostic vector tile base, however, enables multiple styles to be hosted on osm.org and even for users to load their own local styles. This would demonstrate the USP of OSM as compared to Google Maps; provide an opportunity to resolve the perennial tension between osm-carto as a mapper feedback style and new users' desire for something that "just looks nice"; and encourage a healthy ecosystem around vector tile cartography rather than one dominated by a particular company and its authoring tool.

Obviously osm-carto as one (and almost certainly the principal) consumer of these vector tiles would have significant input into the schema. But the schema should be agnostic rather than being designed for the needs of one particular map style.

I'd suggest that a good place to start would be a new experimental project for this vector tiles schema. I'm sure people like @pnorman and @rory would have a lot they could bring to that.

**Comment by kocio-pl** [comment]
Any ideas how such schema should look like to be style-agnostic? I mean how should it be different from just the osm-carto vector tiles?

**Comment by imagico** [comment]
@systemed - in principle i am with you here, back in 2015 i called this [establishing a middle layer between mapping and rendering](http://blog.imagico.de/innovation-and-flexibility-in-the-openstreetmap-project/).

But there is a very strong risk of underestimating the difficulty of actually making this truly style-agnostic, in particular since you - for generating vector tiles - will obviously want to have all the expensive processing steps already in the vector tiles and these expensive parts also tend to be full of subjective design decisions which makes them inherently non-generic and not style-agnostic.

Within the cartographic scope of this style, in particular with the mapper feedback function and the fairly direct and non-complex relationship between original data and rendering result this requires, designing fairly generic vector tiles would be significantly easier - if you accept a significant data volume overhead in the form of the vector tiles containing - for every style they are used for - a significant amount of data that is ultimately not actually used in rendering.

But even then - if you take for example some of the closest variants of this style out there - the German and French style, https://github.com/SomeoneElseOSM/openstreetmap-carto-AJT and maybe also some of the things i implemented in https://github.com/imagico/openstreetmap-carto/tree/alternative-colors - it would already be really difficult to design a vector tile schema that would allow reproducing all of these variants.  In return, if a vector tile schema does not allow for this bandwidth of map design this would severely limit the options you would later have in actual map styling.

But yes, decoupling development of cartographic data processing - you could also call it designing a data schema to be used in rendering - from the actual rendering is something that is highly desirable and quite underdeveloped in the OSM ecosystem so far.

**Comment by andrzej-r** [comment]
I think tiles, whether vector or raster are not a place for generalisation. They are already strongly coupled to projection and zoom levels. If we want them to perform efficiently they should store minimum amount of data required for rendering. If other styles end up needing the same data - great, but we shouldn't dump half of the database into tiles for the sake of flexibility. We would just end up with a less efficient implementation of the database. 

Now, there is nothing stopping us from generating multiple sets of vector tiles, just like we are doing it now with raster ones. The difference is, we could then combine several such tiles during rendering. For example, we could move all labels to separate tiles, one for each language version, or transport routes, or pois.

**Comment by kocio-pl** [comment]
The key problem with serving generic/agnostic vector tiles is hardware and performance.

It's good to have all the data available on-line, but we just have too many of them to do it. For example OSM Carto filters some data while importing into rendering database (see [this part of lua file](https://github.com/gravitystorm/openstreetmap-carto/blob/master/openstreetmap-carto.lua#L50)). Then there are some big SQL queries in `project.mml`, which are optimized to fetch only some data on some zoom levels. Sometimes it would be more readable to write them as Carto CSS rules in MSS files, but that would make everything more resource hungry. And even with all these constrains, tile requests on OSM.org are dropped from a rendering queue on a daily basis - some server resources are not fully used only in the night:

https://munin.openstreetmap.org/openstreetmap/render.openstreetmap/renderd_processed.html

> Now, there is nothing stopping us from generating multiple sets of vector tiles, just like we are doing it now with raster ones.

I like this idea very much. This way or another we need some additional resources for producing vector tiles (it's still very important to find them), but this sounds like a good value for the money. I can imagine that many map styles will need roads, water, buildings, addresses, POIs, admin borders and at least some natural features, so they can have a priority.

**Comment by gravitystorm** [comment]
> Moving to vector tiles gives an opportunity to decouple cartography and schema, which I believe is essential.

I would suggest that this opportunity is deferred, and instead the simplest-possible move to vector tiles is done first. Any decoupling / generalisation / multiple style support should be considered separately as a follow-on project. If too many complex things are attempted simultaneously, then nothing gets completed.

I have experience in designing vector tilesets that are used by multiple styles, and it was hard enough when I had full control over the styles, and the opportunity to work on the system full time too!

I suggest that efforts are put on making the next steps - converting this style to use vector tiles, and deploying to the OSMF hardware - as small and as focussed as possible. Anything that is not strictly necessary can come later.

**Comment by pnorman** [comment]
> Moving to vector tiles gives an opportunity to decouple cartography and schema, which I believe is essential.

I'd like to see this happen. But I also think this is the worst style to do it with, because osm-carto's needs are the most complex. In each of the last 5 minor versions we've made changes which would increment the vector tile schema version. In only one of them was I able to see that it was clearly a simple compatible change.

Thinking about the problem more generally, I don't think it's possible to keep up a good development pace, have a schema used by other renderings that are not closely-tied, and represent the full scope of OSM data. Only two are possible at once.

Tilezen and OpenMapTiles drop some combination of the three, and I consider those the most likely routes for this to happen.

-----

> > There's no rewriting involved. See @rory's talk for what's needed to port the style.
> 
> I will watch it once again, but I'm not sure what your project has to do with his effort. I can understand that vector tile production and serving might be done with different tools, but I don't understand how the CartoCSS rules can be used in Bolder?

Nothing. Rory's effort was about porting a CartoCSS Mapnik style to a CartoCSS tm2style + tm2source style for server-side vector tile rendering. Bolder is a new client-side rendered style, and doesn't use CartoCSS, tm2source, tilelive, node-mapnik, or Mapnik. I don't understand why Bolder keeps getting mentioned here. I'm not reproducing OpenStreetMap Carto technical or cartographically with it.

**Comment by kocio-pl** [comment]
> I don't understand why Bolder keeps getting mentioned here. I'm not reproducing OpenStreetMap Carto technical or cartographically with it.

Oh, now I see. I thought that it was somehow related to your talk with Rory, but it's just a different vector project using the same database schema.

So for now we don't have a software stack for migration and I'm not sure how should it look like. However we can make current style suitable for using vectors as mentioned in https://github.com/gravitystorm/openstreetmap-carto/issues/3201#issuecomment-385417611. Unfortunately this is something which requires more coding skills to start with than I have, so I wait for example changes which I could follow.

**Comment by matthijsmelissen** [comment]
ESRI has created a vector implementation of this style in ArcGIS:

http://esri.maps.arcgis.com/apps/View/index.html?appid=d6b18a2e774c4959ba855f6ac90952a2

Impressive work!

**Comment by kocio-pl** [comment]
The code is here:

https://www.arcgis.com/sharing/rest/content/items/3e1a00aeae81496587988075fe529f71/resources/styles/root.json?f=pjson

Some more informations from developers:

https://github.com/openstreetmap/operations/issues/214#issuecomment-408683834

**Comment by kocio-pl** [comment]
See also #3912.

**Comment by jeisenbe** [comment]
Most of the layers have been switched to us ST_PointOnSurface, or have a PR pending for the switch.

Still remaining are `junctions` and `roads-area-text-name`.

**Comment by dieterdreist** [comment]
is there a solution to unfortunate icon placement like in this case:
https://www.openstreetmap.org/way/22883270#map=13/41.8133/12.2508
The label and icon are placed in a corner rather than close to the access.

**Comment by kocio-pl** [comment]
There is a general and indirect one - to deploy Mapnik current `interior` algorithm to a PostGIS through a GEOS library, which it uses, see https://github.com/gravitystorm/openstreetmap-carto/pull/3874#issuecomment-529172782.

If someone is willing to do that, it would be really great. Do you have an idea who could try to do this?

**Comment by jeisenbe** [comment]
There are now PRs open which will move the remaining layers to use ST_PointOnSurface.

There is also a PR to remove use of classes - #3912 

The next steps for vector files would be:

1) Use ST_Border to filter admin geoms
- See explanation at https://rory.github.io/sotm-2016-talk-vector-tiles/#/20 

2)  Rewrite road layers
- the most difficult step, per @pnorman - I'm not sure what is involved in this step

3) Decide on tools for generating and rendering vector tiles:
 a) Identify a vector tile generator - "perhaps based on ST_AsMVT, e.g PosTile"
 b) Identify a vector to raster tile server - "Tessera should work. All options are painful because they involve node-mapnik"

4) Fork osm-carto and port it to have a source and style directory

5) Set min and max zooms for layers

6) Rewrite way_pixel selectors for z15+
- See https://rory.github.io/sotm-2016-talk-vector-tiles/#/42 for issue

"This would get us a working style and allow us to start exploring what works and doesn't" - per @pnorman

**Comment by jeisenbe** [comment]
Re: step 1. Use ST_Border to filter admin geoms

We could use `ST_Boundary(way) as way` for `admin-text` and `protected-areas-text`. Would it also be needed for `admin-low-zoom`, `admin-mid-zoom` and `admin-high-zoom` or is there no problem since they are from `planet_osm_roads`? 

Could we instead switch to `planet_osm_roads` for the text to get the same effect?

**Comment by jeisenbe** [comment]
Oh, planet_osm_roads doesn't include names so it's not an option, and planet_osm_line does not appear to include administrative boundaries or protected areas, so that's why we need to use `ST_Boundary(way) as way` to produce lines from the polygons in `planet_osm_polygon`

**Comment by jeisenbe** [comment]
@imagico [mentioned](https://github.com/gravitystorm/openstreetmap-carto/pull/3948#issuecomment-545912912):

"Right now the most serious limitations we have in terms of style design are the limitations imposed by the software we use. It seems natural to me that based on that experience we take a critical look at what kind of limitations might be imposed by software we might move to in the future."

"so far there has been no discussion that led to any kind of (even preliminary) decision on concrete software use yet." (https://github.com/gravitystorm/openstreetmap-carto/pull/3948#issuecomment-545935078)

So we should discuss if you are going to move toward server-side vector tiles and how to implement it, and what costs this would impose.

@pnorman previously suggested:

"1. Vector tile generator:

"The options can be divided into ST_AsMVT based or not. I prefer ST_AsMVT, because it avoids a single-company maintainer of the complicated part of vector tile generation.

"PosTile is the leading candidate here, as it supports .tm2source files

"2. Identify a vector to raster tile server

"Tessera should work. All options are painful because they involve node-mapnik."

**Comment by jeisenbe** [comment]
@imagico [mentioned](https://github.com/gravitystorm/openstreetmap-carto/pull/3948#issuecomment-545912912):

"Right now the most serious limitations we have in terms of style design are the limitations imposed by the software we use. It seems natural to me that based on that experience we take a critical look at what kind of limitations might be imposed by software we might move to in the future."

"so far there has been no discussion that led to any kind of (even preliminary) decision on concrete software use yet." (https://github.com/gravitystorm/openstreetmap-carto/pull/3948#issuecomment-545935078)

So we should discuss if you are going to move toward server-side vector tiles and how to implement it, and what costs this would impose.

@pnorman previously suggested:

"1. Vector tile generator:

"The options can be divided into ST_AsMVT based or not. I prefer ST_AsMVT, because it avoids a single-company maintainer of the complicated part of vector tile generation.

"PosTile is the leading candidate here, as it supports .tm2source files

"2. Identify a vector to raster tile server

"Tessera should work. All options are painful because they involve node-mapnik."

**Comment by jeisenbe** [comment]
Is anyone interested in discussing this issue further? 

I opened #3948 to use ST_Boundary, but this is not useful until there is an actual implementation for server-side vector tiles. Do we want to work on this here, or does it need to be a separate repository at this point? 

I would also consider redesigning this raster style for high DPI tiles, which would provide improved style options for thin features like barriers and borders, but I understand than many people are enthusiastic about vector tiles due to the potential to support multiple languages and resolutions.

**Comment by kocio-pl** [comment]
There is a chance to have better rendering in the "middle" of shapes than by `ST_PointOnSurface()`.  In PostGIS 3.1.0alpha2 there's a `ST_MaximumInscribedCircle()` which "finds the largest circle that can by fully contained within a geometry" (a kind of a pole of the inaccesibility), see:

http://postgis.net/2020/07/18/postgis-3.1.0alpha2/
https://download.osgeo.org/postgis/docs/postgis-3.1.0alpha2.pdf (p. 422)

**Comment by verdy-p** [comment]
Could a rendering use a base bitmap layer (without labels/icons, and probably without the linear features with thickness is uncontrolled) so that it can considerably reduce the the complexity for rendering low zoom levels ? This would not require extensive caching as those bitmaps would cover only these low levels and they could be refreshed if needed more often and more precisely (with halftoning and translucent pixels for smoothing curves).
that base layer could probably be using light colors so that detailed vector details (including labels, icons and thick linear features like roads can be visible).
That base layer would gradually become more transparent on higher zoom levels and it's increased pixels smoothed using "mipmapping" technics (the same is those used for rendering 3D scenes with collections of textures at different resolutions smoothed together for almost invisible transitions).

**Comment by verdy-p** [comment]
> There is a chance to have better rendering in the "middle" of shapes than by `ST_PointOnSurface()`. In PostGIS 3.1.0alpha2 there's a `ST_MaximumInscribedCircle()` which "finds the largest circle that can by fully contained within a geometry" (a kind of a pole of the inaccesibility), see:
> 
> http://postgis.net/2020/07/18/postgis-3.1.0alpha2/
> https://download.osgeo.org/postgis/docs/postgis-3.1.0alpha2.pdf (p. 422)

there's also the common transform of 2D shapes into their linear skeleton with nodes connected by straight segments: the node with the longest segment defines not just a maximum radius, and its shortest segment its minimum radius, but it also defines a direction that an be used to properly orient the labels: if you join the segments joining the longest run, you get a line that can also be used to create a spline over which you would orient the label, and the average of the smallest orthogonal segments can be used to scale the fonts...
it could allow correct rendering of shapes without clear borders, using labels or smoothed areas, such as naming mountain ranges, rivers, seas, lakes, natural parks or forests,

**Comment by kocio-pl** [comment]
That would be great to have. Do you know if this feature is implemented in PostGIS?

**Comment by Bibi56** [comment]
@kocio-pl as said in PostGIS 3.1.0alpha2.

**Comment by ZeLonewolf** [comment]
A migration to vector tiles is a complete do-over of the technology stack, even if all the cartographic decisions remain the same.  IMHO that's too big of a change to be executed within this repository (not to mention the name "openstreetmap-carto" no longer makes sense as a repo name -- it's really "openstreetmap-vector" or something along those lines).  A vector version of this style should really start from a new repo and start with some fundamental decisions about what technology stack to build and then iteratively build up the features until they match.

I note the following points which may be of interest to folks looking for an update on vector tile work, and not otherwise mentioned on this thread:
* @pnorman built a partial implementation of the osm-carto style on a vector stack at https://github.com/pnorman/openstreetmap-cartographic ([public demo](https://took.paulnorman.ca/build/mapbox-gl.html)) which could serve as a starting point.
* The OWG under openstreetmap/operations#565 is seeking to stand up a community vector tile server that would support "community style development" which may align with a vector version of this style.

**Comment by jeisenbe** [comment]
> that's too big of a change to be executed within this repository

That is a good point. Should we close this issue, or is there anything still to be done in this style which would be useful for a theoretical vector tile alternative?

**Comment by feludwig** [comment]
Hello, I have been playing around with @pnorman's osm-carto style vector reimplementation, and the whole rendering stack (I recently wrote <https://github.com/feludwig/pgsql-omt-schema>, and a little demo <https://feludwig.github.io/pgsql-omt-schema>)
And I was thinking, could the xml stylesheet not be transpiled to sql+json for client-side vectortiles ?
Assuming client-side rendering (eg with maplibre gl js) of mvt/pbf vector tiles,
two configuration files would be needed :
* json stylesheet client-side
* sql functions server-side (PostGIS has a ST_AsMVT(), and pg_tileserv just serves any sql function that accepts z,x,y coordinates as a vector tiles set)

The sql functions would be easy as a prototype: just take the <DataSource> table from mapnik xml.
- One difficulty is geometry simplification (happens automatically with ST_AsMVTGeom(), but maybe sprinkle some ST_SimplifyPreserveTopology if the mvt tiles are too size-heavy)

The bigger challenge is to transpile all the xml <Rule>s to mapbox-json style layers (there would be a lot of them) and invert the <Min/MaxScaleDenominator> s to "expressions", eg I mean :
```
"icon-color":["step",["zoom"],"blue",15,"red",17,"green"]
```).
- Also needed is a "diff merge" algorithm for "uniting" rules over the same filter, but different fill-colors and any other differing properties
I am playing around with a prototype, but woud there be some gotchas in the mapnik xml <Rule> specification you can think of ?

I looked for the "migration to vector tiles" discussion, is this the correct issue to post into ?

**Comment by imagico** [comment]
> I looked for the "migration to vector tiles" discussion, is this the correct issue to post into ?

This issue was created to discuss ideas and potential concepts to transfer/re-implement this CartoCSS map style for a different tool-chain as an initiative from within the OSM-Carto developer community.  This discussion never went beyond the brainstorming stage, largely because none of the currently available tool-chains for client side rendering is really compatible with this project (both in terms of the cartography and in terms of the openly cooperative development model).

Several independent projects have meanwhile been started to create styles with client side rendering that are designed for some (usually superficial) similarity with OSM-Carto.  The one that got closest so far is actually the earliest (the ESRI demo that was mentioned in the discussion above).

> And I was thinking, could the xml stylesheet not be transpiled to sql+json for client-side vectortiles ?

I am not familiar with projects in that direction.   I don't think this would be practically useful for a map style like OSM-Carto.  Keep in mind this is nothing like the conversion of code between two low level turing-complete languages.  This is about converting map design between two frameworks that are based on completely different overall paradigms.

And this is definitively the wrong forum for asking this kind of question.
